### DISCUSSION TITLE
dealing with async dns request

### DISCUSSION THREAD

**Discussion by barianiluca** [rspamd/rspamd#4664] [General]
Dear,
on one of my custom selectors i'm dealing with the async DNS resolver -> task:get_resolver()
Reading the doc i see that these requests are done in async fashion, so they need a callback where to do the job:
```
-- m['domain'] -> mydomain.com the domain that i want find out mx records
r:resolve_mx(task:get_session(), task:get_mempool(), m['domain'], dns_mx_cb)
-- where dns_mx_cb is my callback
```

the callback works, and i can log the result, but what if i need to use the result of the request immediately on my selector? Because i need to find out if the mail that the user is going to send without auth is for a recipient local or a remote one. So:

```
[...]
if not user then
  if mail_recipients then
    for _,m in pairs(mail_recipients) do
      rspamd_logger.infox("[LDEBUG] localemail -> %s user %s domain %s", m['addr'], m['user'], m['domain'])
      r:resolve_mx(task:get_session(), task:get_mempool(), m['domain'], dns_mx_cb)
      -- HERE I NEED IMMEDIATELY THE RESULT OF THE dns_mx_cb for m['domain'] TO USE ON MY NEXT BUSINESS LOGIC
    end
   [...]
  end
end
[...]
```

There is a way to archive that?
Or this is a wrong approach?
thank you
L.

**Comment by fatalbanana** [comment]
Selectors seem wrong for the task at hand indeed; I would suggest to use a Lua rule.

Relying on DNS also seems unfavourable; I'd suggest to use a local data source (like a map) for this if possible.

~~~lua
-- /etc/rspamd/rspamd.local.lua
-- TODO: consider using local data instead of the DNS or adding caching using Redis

local logger = require 'rspamd_logger'
local lua_util = require 'lua_util'

local our_mx = lua_util.list_to_hash {
  'mail.example.net',
}

local rcpt_mx_id = rspamd_config:register_symbol {
  name = 'RCPT_MX_CHECK',
  type = 'callback',
  callback = function(task)
    local found_local, found_remote, found_error = false, false, false
    local num_waiting = 0

    local cb = function(resolver, to_resolve, results, err)
      num_waiting = num_waiting - 1

      if err then
        logger.err(task, 'lookup error %s: %s', to_resolve, err)
	found_error = true
      else
        for _, result in ipairs(results) do
          if our_mx[result.name] then
            found_local = true
          else
            found_remote = true
          end
        end
      end

      if num_waiting == 0 then
        if found_local then
          task:insert_result('RCPT_MX_LOCAL', 1.0)
	end
        if found_remote then
          task:insert_result('RCPT_MX_REMOTE', 1.0)
	end
        if found_error then
          task:insert_result('RCPT_MX_UNSURE', 1.0)
	end
      end
    end

    local rcpts = task:get_recipients('smtp')
    if not rcpts then return end
    -- get a deduplicated list of recipient domains
    local rcpt_domains_h = {}
    for _, rcpt in ipairs(rcpts) do
      rcpt_domains_h[rcpt.domain] = true
    end
    local rcpt_domains = {}
    for domain in pairs(rcpt_domains_h) do
      table.insert(rcpt_domains, domain)
    end
    -- set a limit on the number of lookups we're willing to do
    rcpt_domains[11] = nil -- up to 10

    for _, domain in ipairs(rcpt_domains) do
      num_waiting = num_waiting + 1
      task:get_resolver():resolve_mx({name = domain, callback = cb})
    end
  end,
}

rspamd_config:register_symbol {
  name = 'RCPT_MX_LOCAL',
  parent = rcpt_mx_id,
  score = 0.0,
  type = 'virtual',
  description = 'RCPT domains with known MX found',
}

rspamd_config:register_symbol {
  name = 'RCPT_MX_REMOTE',
  parent = rcpt_mx_id,
  score = 0.0,
  type = 'virtual',
  description = 'RCPT domains with unknown MX found',
}

rspamd_config:register_symbol {
  name = 'RCPT_MX_UNSURE',
  parent = rcpt_mx_id,
  score = 0.0,
  type = 'virtual',
  description = 'met errors checking RCPT domains MX',
}
~~~

Such a rule might be applied in composites (as `RCPT_MX_LOCAL & !RCPT_MX_REMOTE & !RCPT_MX_UNSURE` for example), or otherwise adapted to suit.

**Comment by barianiluca** [comment]
As usual I come back after a while because i took some times to understand your code. Thank you very much for writing this for me.
I was able to understand it and use it, but I end up with a couple of questions if you want to:

1) my main scope was to limit the non-authenticated email (eg. some users that send the email in old fashion, like `<?php mail("somedestination@somedomains.com","subject","mailbody"); `. This email goes out via sendmail, and not require any auth. I'd like to stop this email to something like 20 per day, in order to force the users to use smtp-auth. So in order to do that, there is a way to create a ratelimit rates that can understand that the user is not authenticated?

```
not_authenticated_mail_daily_limit = {
          #
          # Default 20 mail / 24h for NOT authenticated user
          #
          selector = 'non_authenticated_user_default_selector';
          bucket = [
          {
                burst = 20;
                rate = "20 / 24h";
          }]
        }
```

because the MX check is fine (if the mail is local, is ok, i need to stop only the remote/relay one), but I also need something in order to avoid authenticated user comes into this rate.

2) on my selector I have some trouble to make it to check the symbol created above. I try with something like this:

```
selector = 'symbol("RCPT_MX_LOCAL")'
```

but the rates in not grabbed, but the symbol is correctly registered because i can find it on the log:

```
2023-11-02 16:32:02 #139952(normal) <461f97>; task; rspamd_task_write_log: id: <20231102153202.A02D060C0A@myserver.test>,
qid: <A02D060C0A>, ip: 127.0.0.1, from: <root@myserver.test>, (default: F (no action): [-0.10/10.00] [MIME_GOOD(-0.10){text/plain;},
ARC_NA(0.00){},DBL_FAIL(0.00){localhost:query refused;},FROM_EQ_ENVFROM(0.00){},FROM_HAS_DN(0.00){},MID_RHS_MATCH_FROM(0.00){},MIME_TRACE(0.00){0:+;},
RCPT_COUNT_ONE(0.00){1;},RCPT_MX_LOCAL(0.00){},RCVD_COUNT_ZERO(0.00){0;},SURBL_MULTI_FAIL(0.00){localhost:query refused;},TO_DN_NONE(0.00){},TO_MATCH_ENVRCPT_ALL(0.00){},
URIBL_MULTI_FAIL(0.00){localhost:query refused;}]), len: 245, time: 80.707ms, dns req: 3, digest: <531c376c65c09de0f7152cff1c110538>,
rcpts: <testdev@mydomain.test>, mime_rcpts: <testdev@mydomain.test>, settings_id: ispc_spamfilter_user_3
```

Thank you.

**Reply by fatalbanana** [reply]
> my main scope was to limit the non-authenticated email (eg. some users that send the email in old fashion, like <?php mail("somedestination@somedomains.com","subject","mailbody"); . This email goes out via sendmail, and not require any auth. I'd like to stop this email to something like 20 per day, in order to force the users to use smtp-auth. So in order to do that, there is a way to create a ratelimit rates that can understand that the user is not authenticated?

Depending on integration Rspamd might know which usernames are sending such mails (visible via `task:get_user()` and in logs), if that helps. Do you use Postfix? Not clear to me if/how it would be able to defer or reject such messages but I expect there is a reasonable chance it doesn't work. It might be helpful to use a smarthost for such mail & implement controls there; if headers with identifying information (like a name of a website) are reliably added those could be used for ratelimits, or you might ratelimit the whole server.

> 2. on my selector I have some trouble to make it to check the symbol created above. I try with something like this:

Not clear to me where you try go with it; you could use something like `symbol('RCPT_MX_LOCAL').id('my-local-mx-limit')` to limit all messages with this symbol to a common rate but I suspect it's not what you want.

**Reply by barianiluca** [reply]
ok probably i was be able to solve in a quicker and simpler way. The MX check was way too complicated.
Due that, at the end of the day what I want was to limit the non-authenticated mail to 20 by day (so the mail sent with php via sendmail), i archived that in that way:

- set php to put the header `X-PHP-Originating-Script` with the config `mail.add_x_header = On` (this setting put that header for all the non authenticated email sent via php;
- then from there i have create a new selector that check these email, and count them using the UID passed on the headers by php:
```
-- /etc/rspamd/rspamd.local.lua
[...]
lua_selectors.register_extractor(rspamd_config, "uid_0_per_day_limit_selector", {
  get_value = function(task, args)
    local x_php_originating_script = task:get_header("X-PHP-Originating-Script")
    if not x_php_originating_script then return end
    local uid = x_php_originating_script:match("^[0-9]+")
    if not uid then return end
    if uid_0_per_day_limit:get_key(uid) then
        rspamd_logger.infox("Blocked website for spam - UID %s", uid)
        return
    end
    return uid, 'string'
  end,
  description = 'Website with spam problem. Blocked to avoid spammy traffic'
})
```
```
-- /etc/rspamd/local.d/ratelimit.conf
[...]
not_authenticated_mail_daily_limit = {
          #
          # Default 20 mail / 24h for NOT authenticated user
          #
          selector = 'uid_0_per_day_limit_selector';
          bucket = [
          {
                burst = 20;
                rate = "20 / 24h";
          }]
        }
```

This seems to work pretty well.
Thank you for all the support.
